src/main/java/org/jooq/lambda/SeqBuffer.java

```diff
@@ -0,0 +1,110 @@
+package org.jooq.lambda;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * Lazily consumes given <code>Spliterator</code> through <code>Seq</code>s provided by method <code>seq()</code>.
+ * This method may be called multiple times, and the returned <code>Seq</code>s may be consumed interchangeably.
+ *
+ * This class is not thread-safe.
+ *
+ * @author Tomasz Linkowski
+ */
+final class SeqBuffer<T> {
+
+    static <T> SeqBuffer<T> of(Stream<T> stream) {
+        return of(stream.spliterator());
+    }
+
+    static <T> SeqBuffer<T> of(Spliterator<T> spliterator) {
+        if (spliterator instanceof SeqBuffer.BufferSpliterator) {
+            return ((SeqBuffer<T>.BufferSpliterator) spliterator).parentSeqBuffer(); // reuse existing SeqBuffer
+        }
+        return new SeqBuffer<>(spliterator);
+    }
+
+    private final Spliterator<T> source;
+    private final List<T> buffer = new ArrayList<>();
```

Comment:
Hmm, this buffer is not thread safe. It may well be that one consumer of a buffered result consumes the buffer while another consumer produces new elements at the same time. E.g., this could happen in your `splitAt()` implementation suggestion here: https://github.com/jOOQ/jOOL/pull/305/commits/b345bbb8244e727d01b7a59ee10835723e5da72c